### PR TITLE
Js udf scalar

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzedStatement.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+package io.crate.analyze;
+
+import io.crate.sql.tree.Expression;
+import io.crate.types.DataType;
+
+import java.util.List;
+
+public class CreateFunctionAnalyzedStatement extends AbstractDDLAnalyzedStatement {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzedStatement.java]: The base class doesn't really fit as this statement doesn't have tableParameters

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import io.crate.metadata.Schemas;
+import io.crate.sql.tree.CreateFunction;
+import io.crate.sql.tree.FunctionArgument;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CreateFunctionAnalyzer {
+
+    private final static DataTypeAnalyzer DT_ANALYZER = new DataTypeAnalyzer();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java]: I think we could make DataTypeAnalyzer a singleton as it doesn't have any state

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            if (context.sessionContext().defaultSchema() != null) {
+                return context.sessionContext().defaultSchema();
+            }
+            return Schemas.DEFAULT_SCHEMA_NAME;
+        } else {
+            return parts.get(0);
+        }
+    }
+
+    private String resolveFunctionName(List<String> parts) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java]: can be static

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            resolveFunctionName(parts),
+            node.replace(),
+            analyzedFunctionArguments(node.arguments()),
+            DT_ANALYZER.process(node.returnType(), null),
+            node.language(),
+            node.definition()
+        );
+    }
+
+    private String resolveSchemaName(List<String> parts, Analysis context) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java]: can be static and I'd change the second param to defaultSchema directly as that's the only thing out of the analysis which is needed

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        } else {
+            return parts.get(0);
+        }
+    }
+
+    private String resolveFunctionName(List<String> parts) {
+        return parts.size() == 1 ? parts.get(0) : parts.get(1);
+    }
+
+    private List<FunctionArgumentDefinition> analyzedFunctionArguments(List<FunctionArgument> arguments) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java]: can be static

File: sql/src/main/java/io/crate/analyze/DropFunctionAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.stream.Collectors;
+
+class DropFunctionAnalyzer {
+
+    private final static DataTypeAnalyzer DT_ANALYZER = new DataTypeAnalyzer();
+
+    private List<FunctionArgumentDefinition> analyzedFunctionArguments(List<FunctionArgument> arguments) {
+        return arguments.stream()
+            .map(arg ->
+                FunctionArgumentDefinition.of(arg.name().orElse(null), DT_ANALYZER.process(arg.type(), null))
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/DropFunctionAnalyzer.java]: Maybe `FunctionArgumentDefinition` could have a static factory method which does this `List<FunctionArgument>` -> `List<FunctionArgumentDefinition>` conversion. This method is the same as in the CreateFunctionAnalyzer.

File: sql/src/main/java/io/crate/analyze/FunctionArgumentDefinition.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        return new FunctionArgumentDefinition(null, dataType);
+    }
+
+    public static FunctionArgumentDefinition fromStream(StreamInput in) throws IOException {
+        FunctionArgumentDefinition argumentDefinition = new FunctionArgumentDefinition();
+        argumentDefinition.readFrom(in);
+        return argumentDefinition;
+    }
+
+    public void name(String name) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/FunctionArgumentDefinition.java]: why does this need setters?

File: sql/src/main/java/io/crate/analyze/FunctionArgumentDefinition.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+public class FunctionArgumentDefinition implements Streamable, ToXContent {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/FunctionArgumentDefinition.java]: why do we need ToXContent here?

File: sql/src/main/java/io/crate/operation/udf/CreateUserDefinedFunctionAction.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package io.crate.operation.udf;
+
+
+import org.elasticsearch.action.Action;
+import org.elasticsearch.client.ElasticsearchClient;
+
+public class CreateUserDefinedFunctionAction extends
+    Action<CreateUserDefinedFunctionRequest, CreateUserDefinedFunctionResponse, CreateUserDefinedFunctionRequest.RequestBuilder> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/CreateUserDefinedFunctionAction.java]: Action should only be required to support the ES TransportClient which we don't need to support for internals. I think this can be removed

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.udf;
+
+public class UserDefinedFunction {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java]: what is this for?

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionDDLDispatcher.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+import java.util.concurrent.CompletableFuture;
+
+@Singleton
+public class UserDefinedFunctionDDLDispatcher {
+
+    private final TransportActionProvider transportActionProvider;
+
+    @Inject
+    public UserDefinedFunctionDDLDispatcher(TransportActionProvider transportActionProvider) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionDDLDispatcher.java]: Is it possible to inject `TransportCreateUserDefinedFunctionAction` directly here?

File: sql/src/main/java/io/crate/exceptions/UserDefinedFunctionAlreadyExistsException.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+public class UserDefinedFunctionAlreadyExistsException extends ConflictException {
+
+    public UserDefinedFunctionAlreadyExistsException(UserDefinedFunctionMetaData udfMetaData) {
+        super(String.format(Locale.ENGLISH, "User defined Function '%s(%s)' already exists.",
+            udfMetaData.name(), Joiner.on(", ").join(
+                udfMetaData.arguments().stream().map(FunctionArgumentDefinition::type).collect(Collectors.toList())
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/exceptions/UserDefinedFunctionAlreadyExistsException.java]: Could use `Collectors.joining`

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 13]
```diff
@@ -31,73 +31,116 @@
...
         Multimap<String, Tuple<FunctionIdent, FunctionImplementation>> signatureMap = ArrayListMultimap.create();
         for (Map.Entry<FunctionIdent, FunctionImplementation> entry : functionImplementations.entrySet()) {
             signatureMap.put(entry.getKey().name(), new Tuple<>(entry.getKey(), entry.getValue()));
         }
-        for (String name : signatureMap.keys()) {
-            functionResolvers.put(name, new GeneratedFunctionResolver(signatureMap.get(name)));
+        return signatureMap;
+    }
+
+    public void registerSchemaFunctionResolvers(String schema, Map<String, FunctionResolver> resolvers) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: doesn't this have to be thread-safe ?

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 14]
```diff
@@ -31,73 +31,116 @@
...
+        if (schema == null) {
+            FunctionResolver dynamicResolver = functionResolvers.get(name);
+            if (dynamicResolver != null) {
+                List<DataType> signature = dynamicResolver.getSignature(arguments);
+                if (signature != null) {
+                    return dynamicResolver.getForTypes(signature);
+                }
+            }
+            // fallback, check if the function with the same name was defined in the default schema
+            return getSchemaFunctionImplementation(Schemas.DEFAULT_SCHEMA_NAME, name, arguments);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: I thought this fallback *shouldn't* be done? Can you please describe the lookup-order in the Functions class javadocs

File: sql/src/main/java/io/crate/operation/scalar/arithmetic/FloorFunction.java [hunk 15]
```diff
@@ -97,6 +97,9 @@ public FunctionInfo info() {
 
         @Override
         public Number evaluate(Input<Number>[] args) {
+            if (args == null) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/arithmetic/FloorFunction.java]: This case should never happen, why is this now necessary?

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+package io.crate.operation.udf;
+
+import io.crate.data.Input;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.types.DataType;
+
+public abstract class UserDefinedFunction extends Scalar<Object, Object> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java]: I think we should remove this  and instead of Scalar directly

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @SafeVarargs
+    @Override
+    public final Object evaluate(Input<Object>... args) {
+        Object[] values = new Object[args.length];
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunction.java]: Please avoid the object array creation and instead handle the inputs directly in the concrete implementations. This just leads to inefficient implementations

**Comment by mfussenegger** [comment]
tests are broken?

File: sql/src/main/java/io/crate/metadata/Functions.java (hunk 14 above)
Comment (kovrus) [inline sql/src/main/java/io/crate/metadata/Functions.java]: removed

**Comment by kovrus** [comment]
itest are broken, but will it is going to be fixed wuth the nexttwo prs

File: sql/src/main/java/io/crate/analyze/DataTypeAnalyzer.java [hunk 18]
```diff
@@ -33,6 +33,8 @@
 
 public class DataTypeAnalyzer extends DefaultTraversalVisitor<DataType, Void> {
 
+    public final static DataTypeAnalyzer INSTANCE = new DataTypeAnalyzer();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/DataTypeAnalyzer.java]: please also add a private ctor to enforce using the singleton. 

Instead of exposing the INSTANCE it would also be possible to have a static `DataType convert(Expression expr)` method

File: sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            resolveFunctionName(parts),
+            node.replace(),
+            toFunctionArgumentDefinitions(node.arguments()),
+            DataTypeAnalyzer.INSTANCE.process(node.returnType(), null),
+            node.language(),
+            node.definition()
+        );
+    }
+
+    private static String resolveSchemaName(List<String> parts, Analysis context) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/CreateFunctionAnalyzer.java]: please replace the Analysis param with just the schema name

File: sql/src/main/java/io/crate/metadata/FunctionIdent.java [hunk 20]
```diff
@@ -38,30 +36,41 @@
...
     private String name;
     private List<DataType> argumentTypes;
 
     public FunctionIdent() {
-
     }
 
-    public static FunctionIdent of(String name, DataType type1, DataType type2) {
-        return new FunctionIdent(name, ImmutableList.of(type1, type2));
+    public FunctionIdent(String schema, String name, List<DataType> argumentTypes) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/FunctionIdent.java]: please mark schema as nullable

File: sql/src/test/java/io/crate/operation/aggregation/impl/StdDevAggregationTest.java [hunk 21]
```diff
@@ -40,9 +40,10 @@
     @Test
     public void testReturnType() throws Exception {
         for (DataType<?> type : Iterables.concat(DataTypes.NUMERIC_PRIMITIVE_TYPES, Arrays.asList(DataTypes.TIMESTAMP))) {
-            FunctionIdent fi = new FunctionIdent("stddev", ImmutableList.<DataType>of(type));
+            FunctionIdent fi = new FunctionIdent();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/aggregation/impl/StdDevAggregationTest.java]: FunctionIdent is now unused?

File: sql/src/test/java/io/crate/operation/aggregation/impl/MaximumAggregationTest.java [hunk 22]
```diff
@@ -86,6 +84,6 @@ public void testString() throws Exception {
 
     @Test(expected = NullPointerException.class)
     public void testUnsupportedType() throws Exception {
-        Object[][] result = executeAggregation(DataTypes.OBJECT, new Object[][]{{new Object()}});
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/aggregation/impl/MaximumAggregationTest.java]: Please avoid unrelated cleanups like this in a feature PR like this. Makes it much more difficult to review

File: sql/src/test/java/io/crate/metadata/FunctionsTest.java [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+            new SessionContext(0, Option.NONE, "bar"),
+            new ParameterContext(Row.EMPTY, Collections.emptyList())).analyzedStatement();
+        assertThat(analysis.relation().querySpec().outputs().get(0), isLiteral(1L));
+    }
+
+    @Test
+    public void testUserDefinedFunctionInCustomSchemaAccessibleViaSchemaName() throws Exception {
+        SelectAnalyzedStatement analysis = (SelectAnalyzedStatement) e.analyzer.boundAnalyze(
+            SqlParser.createStatement("select custom.foo('foo');"),
+            new SessionContext(0, Option.NONE, null),
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/metadata/FunctionsTest.java]: please extract all this boilerplate into a method or something to reduce the noise per test

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+import org.junit.Test;
+
+import java.util.HashMap;
+
+import static org.hamcrest.CoreMatchers.is;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 2, numClientNodes = 0, randomDynamicTemplates = false)
+public class UserDefinedFunctionsIntegrationTest extends SQLTransportIntegrationTest {
+
+    private final Object[][] rows = new Object[][]{
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: please avoid using final in tests - we've a memory-leak detection that will start failing if tests keep too many things as final

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+        execute("create table test " +
+            "(id long, str string, obj object, arr array(long)) " +
+            "clustered by(id) into 2 shards");
+
+        execute("insert into test (id, str, obj, arr) values (?, ?, ?, ?)", rows);
+        refresh();
+    }
+
+    @Test
+    public void testCreateFunctionWithObjectInputType() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: Is it possible to remove some of the integration tests here and replace them with unit tests? To me it's unclear if a test is about storing the function, or about the lookup, both or something else entirely.

File: sql/src/test/java/io/crate/integrationtests/PostgresITest.java [hunk 26]
```diff
@@ -633,6 +633,18 @@ public void testSetMultipleSchemasOnSession() throws Exception {
         }
     }
 
+    @Test
+    public void testUseBuiltInFunctionWithSetSchemaOnSession() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: PostgresITest should be kept to postgres protocol specific tests - don't abuse it to set a custom schema.

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionsMetaData.java [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+    public static final String TYPE = "user_defined_functions";
+
+    public static final UserDefinedFunctionsMetaData PROTO = new UserDefinedFunctionsMetaData();
+
+    static {
+        // register non plugin custom metadata
+        MetaData.registerPrototype(TYPE, PROTO);
+    }
+
+    private final Map<Integer, UserDefinedFunctionMetaData> functions;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionsMetaData.java]: please rename to indicate what the key is - `functionsByXY` or something

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+            .custom(UserDefinedFunctionsMetaData.TYPE);
+
+        // Check if user defined functions got changed
+        if ((oldMetaData == null && newMetaData == null) || (oldMetaData != null && oldMetaData.equals(newMetaData))) {
+            return;
+        }
+
+        Map<String, Map<FunctionIdent, FunctionImplementation>> schemaFunctions = newMetaData.functionsMetaData().stream()
+            .map(UserDefinedFunctionFactory::of)
+            .collect(groupingBy(f -> f.info().ident().schema(), toMap(f -> f.info().ident(), f -> f)));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java]: couldn't the loop be changed to go through the list instead? I think the map creation is not really necessary

File: sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+            functions.registerSchemaFunctionResolvers(
+                entry.getKey(),
+                functions.generateFunctionResolvers(entry.getValue())
+            );
+        }
+    }
+
+    static class RegisterUserDefinedFunctionRequest extends ClusterStateUpdateRequest<RegisterUserDefinedFunctionRequest> {
+
+        UserDefinedFunctionMetaData metaData;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java]: can't metaData also be final?

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 30]
```diff
@@ -31,73 +31,123 @@
...
+    }
+
+    public void registerSchemaFunctionResolvers(String schema, Map<String, FunctionResolver> resolvers) {
+        schemaFunctionResolvers.put(schema, resolvers);
+    }
+
+    /**
+     * Returns the function implementation for the given schema, function name and arguments.
+     *
+     * If the explicit schema name is not provided then the first function look-up will be done
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: Maybe rewrite this as example:

```
x.foo()     -> lookup in x
foo()        -> lookup in built-ins, fallback lookup in session schema
```

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 31]
```diff
@@ -31,73 +31,123 @@
...
+        if (schema == null) {
+            FunctionResolver dynamicResolver = functionResolvers.get(name);
+            if (dynamicResolver != null) {
+                List<DataType> signature = dynamicResolver.getSignature(arguments);
+                if (signature != null) {
+                    return dynamicResolver.getForTypes(signature);
+                }
+            }
+            // check if the function without explicit schema name exists in the default schema
+            return getSchemaFunctionImplementation(Schemas.DEFAULT_SCHEMA_NAME, name, arguments);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: Doesn't this have to fallback to sessionContext.defaultSchema (and only if that is false, use DOC)

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java (hunk 25 above)
Comment (kovrus) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: i will keep only one integration test for overlaoding, all other cases are tested in unit tests.

File: sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+
+import javax.script.*;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class JavaScriptUserDefinedFunction extends Scalar<Object, Object> {
+
+    static final ScriptEngine ENGINE = new ScriptEngineManager().getEngineByName("nashorn");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java]: Is nashorn available on all JVMs? How would it behave if not?

File: sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+            result = parseScriptObject((ScriptObjectMirror) result);
+        }
+        return returnType.value(result);
+    }
+
+    private Object parseScriptObject(ScriptObjectMirror scriptObject) {
+        switch (returnType.id()) {
+            case ArrayType.ID:
+                if (scriptObject.isArray()) {
+                    return scriptObject.values().stream().toArray(Object[]::new);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java]: Could call .toArray() directly on values() without using stream

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+        new Object[]{2L, "bar", new HashMap<String, Long>() {{
+            put("foo", 2L);
+        }}, new Object[]{2L, 3L}
+        }
+    };
+
+    @Before
+    public void beforeTest() {
+        execute("create table test " +
+            "(id long, str string, obj object, arr array(long)) " +
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: looks like id / clustered by is not relevant for the tests? If so, remove it please

File: sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java (hunk 32 above)
Comment (kovrus) [inline sql/src/main/java/io/crate/operation/udf/JavaScriptUserDefinedFunction.java]: yes, it is available on all jvms

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java (hunk 34 above)
Comment (kovrus) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: I want to guarantee that these two records are in two nodes and that function hits each of them.

Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: This doesn't do what you want then -  1L and 2L seem to be routed to the same shards if there are two 

```
cr> select id, schema_name, num_docs, table_name from sys.shards;
+----+-------------+----------+------------+
| id | schema_name | num_docs | table_name |
+----+-------------+----------+------------+
|  0 | doc         |        0 | t1         |
|  1 | doc         |        2 | t1         |
|  0 | doc         |        0 | t1         |
|  1 | doc         |        0 | t1         |
+----+-------------+----------+------------+
SELECT 4 rows in set (0.003 sec)
cr> select * from t1;
+----+
| id |
+----+
|  1 |
|  2 |
+----+
SELECT 2 rows in set (0.040 sec)
```
